File: packages/provide-uterm-go/shell/dispatcher.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package shell

import (
	"context"
	"net/http"
	"sort"
	"strings"

	"github.com/provide-io/provide-uterm/packages/provide-uterm-go/render"
)

// CommandDispatcher parses and dispatches ushell command lines. Port of
// commands/dispatcher.py:CommandDispatcher.
type CommandDispatcher struct {
	ctx         *Context
	client      *http.Client
	renderImage imageRenderer
}

// NewCommandDispatcher builds a dispatcher over ctx (nil is treated as an empty
// context). It uses http.DefaultClient for fetch/cast/render and the render
// package's image decoder; both are overridable in tests within the package.
func NewCommandDispatcher(ctx *Context) *CommandDispatcher {
	if ctx == nil {
		ctx = &Context{}
	}
	return &CommandDispatcher{
		ctx:         ctx,
		client:      http.DefaultClient,
		renderImage: render.ImageToANSIFrames,
	}
}

// Dispatch processes a completed line and returns the command Result. Port of
// CommandDispatcher.dispatch.
func (d *CommandDispatcher) Dispatch(ctx context.Context, line string) Result {
	line = pyStrip(line)
	// Ctrl+C — already echoed; just re-show the prompt.
	if line == "" || line == "\x03" {
		return textResult(Prompt)
	}

	parts := pySplit1(line)
	cmd := strings.ToLower(parts[0])
	arg := ""
	if len(parts) > 1 {
		arg = pyStrip(parts[1])
	}

	switch cmd {
	case "exit", "quit", "\x04":
		return textResult(InfoMsg("Goodbye.\r\n") + Prompt)

	case "help":
		if arg != "" {
			detail, found := commandHelp[strings.ToLower(arg)]
			if !found {
				return textResult(ErrorMsg("no help for '"+arg+"'") + Prompt)
			}
			return textResult(detail + Prompt)
		}
		return textResult(helpText + Prompt)

	case "clear":
		return textResult(ClearScreen + Prompt)

	case "py":
		return cmdPy(arg)

	case "sessions":
		if strings.HasPrefix(arg, "kill ") || arg == "kill" {
			id := ""
			if strings.HasPrefix(arg, "kill ") {
				id = pyStrip(arg[5:])
			}
			return cmdSessionsKill(ctx, d.ctx, id)
		}
		return cmdSessions(ctx, d.ctx)

	case "kv":
		return cmdKV(ctx, d.ctx, arg)

	case "fetch":
		return cmdFetch(ctx, d.client, arg)

	case "storage":
		return cmdStorage(ctx, d.ctx, arg)

	case "env":
		return d.cmdEnv()

	case "render":
		return cmdRender(ctx, d.client, d.renderImage, arg)

	case "cast":
		return cmdCast(ctx, d.client, arg)
	}

	return textResult(ErrorMsg("unknown command: '"+cmd+"' — type "+Bold+"help"+Reset) + Prompt)
}

// cmdEnv shows available context keys (or env attributes). Port of
// CommandDispatcher._cmd_env.
func (d *CommandDispatcher) cmdEnv() Result {
	var lines []string
	if d.ctx.Env != nil {
		attrs := d.ctx.Env.Attrs()
		names := make([]string, 0, len(attrs))
		for name := range attrs {
			names = append(names, name)
		}
		sort.Strings(names)
		for _, name := range names {
			lines = append(lines, FmtKVDefault(name, attrs[name]))
		}
	} else {
		keys := make([]string, 0, len(d.ctx.Values))
		for k := range d.ctx.Values {
			if !strings.HasPrefix(k, "_") {
				keys = append(keys, k)
			}
		}
		sort.Strings(keys)
		for _, k := range keys {
			lines = append(lines, FmtKVDefault(k, ""))
		}
	}
	output := InfoMsg("(empty context)")
	if len(lines) > 0 {
		output = Heading("context") + strings.Join(lines, "")
	}
	return textResult(output + Prompt)
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Shell/Dispatcher.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Net.Http;

namespace Provide.Uterm.Shell;

/// <summary>
/// Parses and dispatches ushell command lines.
/// Port of packages/provide-uterm-go/shell/dispatcher.go.
/// </summary>
public sealed class CommandDispatcher
{
    private readonly ShellContext _ctx;
    private readonly HttpClient _client;
    private readonly Func<byte[], int, int, string, (IReadOnlyList<string> Frames, double Fps)>? _renderImage;
    private readonly bool _ownsClient;

    public CommandDispatcher(
        ShellContext? ctx = null,
        HttpClient? client = null,
        Func<byte[], int, int, string, (IReadOnlyList<string> Frames, double Fps)>? renderImage = null)
    {
        _ctx = ctx ?? new ShellContext();
        _ownsClient = client is null;
        _client = client ?? new HttpClient();
        _renderImage = renderImage;
    }

    /// <summary>Synchronous dispatch for simple commands (tests / line-oriented callers).</summary>
    public ShellResult Dispatch(string line) =>
        DispatchAsync(line).GetAwaiter().GetResult();

    public async Task<ShellResult> DispatchAsync(string line, CancellationToken ct = default)
    {
        line = StrUtil.PyStrip(line);
        if (line.Length == 0 || line == "\x03")
        {
            return ShellResult.OfText(ShellOutput.Prompt);
        }

        var parts = StrUtil.PySplit1(line);
        if (parts is null || parts.Length == 0)
        {
            return ShellResult.OfText(ShellOutput.Prompt);
        }

        var cmd = parts[0].ToLowerInvariant();
        var arg = parts.Length > 1 ? StrUtil.PyStrip(parts[1]) : "";

        switch (cmd)
        {
            case "exit":
            case "quit":
            case "\x04":
                return ShellResult.OfText(ShellOutput.InfoMsg("Goodbye.\r\n") + ShellOutput.Prompt);

            case "help":
                if (arg.Length > 0)
                {
                    if (!Help.CommandHelp.TryGetValue(arg.ToLowerInvariant(), out var detail))
                    {
                        return ShellResult.OfText(ShellOutput.ErrorMsg("no help for '" + arg + "'") + ShellOutput.Prompt);
                    }

                    return ShellResult.OfText(detail + ShellOutput.Prompt);
                }

                return ShellResult.OfText(Help.HelpText + ShellOutput.Prompt);

            case "clear":
                return ShellResult.OfText(ShellOutput.ClearScreen + ShellOutput.Prompt);

            case "py":
                return Commands.CmdPy(arg);

            case "sessions":
                if (arg.StartsWith("kill ", StringComparison.Ordinal) || arg == "kill")
                {
                    var id = arg.StartsWith("kill ", StringComparison.Ordinal) ? StrUtil.PyStrip(arg[5..]) : "";
                    return await Commands.CmdSessionsKillAsync(_ctx, id, ct).ConfigureAwait(false);
                }

                return await Commands.CmdSessionsAsync(_ctx, ct).ConfigureAwait(false);

            case "kv":
                return await Commands.CmdKvAsync(_ctx, arg, ct).ConfigureAwait(false);

            case "fetch":
                return await Commands.CmdFetchAsync(_client, arg, ct).ConfigureAwait(false);

            case "storage":
                return await Commands.CmdStorageAsync(_ctx, arg, ct).ConfigureAwait(false);

            case "env":
                return CmdEnv();

            case "render":
                return await Commands.CmdRenderAsync(_client, _renderImage, arg, ct).ConfigureAwait(false);

            case "cast":
                return await Commands.CmdCastAsync(_client, arg, ct).ConfigureAwait(false);

            default:
                return ShellResult.OfText(
                    ShellOutput.ErrorMsg("unknown command: '" + cmd + "' — type " + ShellOutput.Bold + "help" + ShellOutput.Reset) +
                    ShellOutput.Prompt);
        }
    }

    private ShellResult CmdEnv()
    {
        var lines = new List<string>();
        if (_ctx.Env is not null)
        {
            var attrs = _ctx.Env.Attrs();
            foreach (var name in attrs.Keys.OrderBy(k => k, StringComparer.Ordinal))
            {
                lines.Add(ShellOutput.FmtKvDefault(name, attrs[name]));
            }
        }
        else
        {
            foreach (var k in _ctx.Values.Keys.Where(k => !k.StartsWith('_')).OrderBy(k => k, StringComparer.Ordinal))
            {
                lines.Add(ShellOutput.FmtKvDefault(k, ""));
            }
        }

        var output = lines.Count > 0
            ? ShellOutput.Heading("context") + string.Join("", lines)
            : ShellOutput.InfoMsg("(empty context)");
        return ShellResult.OfText(output + ShellOutput.Prompt);
    }
}
